File: src/main/scala/com/databricks/labs/validation/utils/Structures.scala
package com.databricks.labs.validation.utils

import com.databricks.labs.validation.Rule
import org.apache.spark.sql.{Column, DataFrame}

object Structures {

  case class Bounds(
                     lower: Double = Double.NegativeInfinity,
                     upper: Double = Double.PositiveInfinity,
                     lowerInclusive: Boolean = false,
                     upperInclusive: Boolean = false) {
    def validationLogic(c: Column): Column = {
      val lowerLogic = if (lowerInclusive) c >= lower else c > lower
      val upperLogic = if (upperInclusive) c <= upper else c < upper
      lowerLogic && upperLogic
    }
  }

  case class MinMaxRuleDef(ruleName: String, column: Column, bounds: Bounds, by: Column*)

  case class ValidationResults(completeReport: DataFrame, summaryReport: DataFrame)

  private[validation] class ValidationException(s: String) extends Exception(s) {}

  private[validation] class InvalidRuleException(r: Rule, s: String) extends Exception(s) {
    val msg: String = s"RULE VALIDATION FAILED: ${r.toString}"
    throw new ValidationException(msg)
  }

}

File: python/databricks/labs/validation/structures.py
import pyspark
from typing import List

from databricks.labs.validation.local_spark_singleton import SparkSingleton


class Bounds:

    def __init__(self, lower, upper,
                 lower_inclusive=False,
                 upper_inclusive=False):
        self.lower = lower
        self.upper = upper
        self.lower_inclusive = lower_inclusive
        self.upper_inclusive = upper_inclusive
        self._spark = SparkSingleton.get_instance()
        self._jBounds = self._spark._jvm.com.databricks.labs.validation.utils.Structures.Bounds(lower, upper,
                                                                                                lower_inclusive,
                                                                                                upper_inclusive)

    def validation_logic(self, col):
        jCol = col._jc
        return self._spark._jvm.com.databricks.labs.validation.utils.Structures.Bounds.validation_logic(jCol)


class MinMaxRuleDef:

    def __init__(self,
                 rule_name: str,
                 column: pyspark.sql.Column,
                 bounds: Bounds,
                 by: List[pyspark.sql.Column] = None):
        self.rule_name = rule_name
        self.column = column
        self.bounds = bounds
        self.by = by


class ValidationResults:

    def __init__(self,
                 complete_report: pyspark.sql.DataFrame,
                 summary_report: pyspark.sql.DataFrame):
        self.complete_report = complete_report
        self.summary_report = summary_report
